File: java/src/main/java/com/os/cvcamera/java/MainActivity.java
package com.os.cvcamera.java;

import android.os.Bundle;
import android.util.Log;
import android.view.WindowManager;
import android.widget.Toast;

import java.util.Collections;
import java.util.List;

import org.opencv.android.CameraActivity;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame;
import org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Mat;
import org.opencv.imgproc.Imgproc;

/**
 * OpenCV example written entirely in Java, with no Kotlin and no native code.
 *
 * Frames are processed through the OpenCV Java bindings. Tapping the preview cycles the effect.
 *
 * This is the counterpart to the Kotlin example in {@code app/} and the C++ example in
 * {@code cpp/}.
 */
public class MainActivity extends CameraActivity implements CvCameraViewListener2 {

    private static final String TAG = "CvCameraJava";

    static {
        System.loadLibrary("opencv_java4");
    }

    /** Effect applied to each frame. */
    private enum Mode {
        NONE,
        GREYSCALE,
        CANNY
    }

    private CameraBridgeViewBase cameraView;
    private Mode mode = Mode.CANNY;

    /** Reused across frames so a new Mat is not allocated per frame. */
    private Mat grey;
    private Mat edges;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        setContentView(R.layout.activity_main);

        Log.d(TAG, "OpenCV version: " + OpenCVLoader.OPENCV_VERSION);

        cameraView = findViewById(R.id.cameraView);
        cameraView.setCvCameraViewListener(this);
        cameraView.setOnClickListener(view -> cycleMode());
    }

    private void cycleMode() {
        switch (mode) {
            case CANNY:
                mode = Mode.GREYSCALE;
                break;
            case GREYSCALE:
                mode = Mode.NONE;
                break;
            default:
                mode = Mode.CANNY;
                break;
        }
        Toast.makeText(this, getString(R.string.mode_changed, mode.name()), Toast.LENGTH_SHORT).show();
    }

    @Override
    protected List<? extends CameraBridgeViewBase> getCameraViewList() {
        return Collections.singletonList(cameraView);
    }

    @Override
    protected void onCameraPermissionGranted() {
        super.onCameraPermissionGranted();
        cameraView.enableView();
    }

    @Override
    public void onCameraViewStarted(int width, int height) {
        grey = new Mat();
        edges = new Mat();
    }

    @Override
    public void onCameraViewStopped() {
        if (grey != null) {
            grey.release();
        }
        if (edges != null) {
            edges.release();
        }
    }

    @Override
    public Mat onCameraFrame(CvCameraViewFrame inputFrame) {
        Mat rgba = inputFrame.rgba();
        switch (mode) {
            case CANNY:
                // Camera frames are CV_8UC4; RGBA2GRAY accepts the alpha channel.
                Imgproc.cvtColor(rgba, grey, Imgproc.COLOR_RGBA2GRAY);
                Imgproc.Canny(grey, edges, 60, 160);
                Imgproc.cvtColor(edges, rgba, Imgproc.COLOR_GRAY2RGBA);
                break;
            case GREYSCALE:
                Imgproc.cvtColor(rgba, grey, Imgproc.COLOR_RGBA2GRAY);
                Imgproc.cvtColor(grey, rgba, Imgproc.COLOR_GRAY2RGBA);
                break;
            default:
                break;
        }
        return rgba;
    }

    @Override
    protected void onPause() {
        super.onPause();
        cameraView.disableView();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cameraView.disableView();
    }
}

File: app/src/main/java/com/al/cvcamera/MainActivity.kt
package com.al.cvcamera


import android.os.Bundle
import android.util.Log
import android.view.SurfaceView
import android.view.Window
import android.view.WindowManager
import com.al.cvcamera.databinding.ActivityMainBinding
import org.opencv.android.*
import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.imgproc.Imgproc


class MainActivity : CameraActivity() ,CameraBridgeViewBase.CvCameraViewListener2 {

    val TAG: String = javaClass.simpleName
    private lateinit var binding: ActivityMainBinding
    private lateinit var mOpenCvCameraView: JavaCamera2View
    private lateinit var mRGBA:Mat
    private lateinit var mRGBAT:Mat

    private val mLoaderCallback: BaseLoaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            when (status) {
                SUCCESS -> {
                    Log.i(TAG, "OpenCV loaded successfully")
                    mOpenCvCameraView.enableView()
                }
                else -> {
                    super.onManagerConnected(status)
                }
            }
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        //OpenCV Camera
        mOpenCvCameraView = binding.CvCamera
        Log.d(TAG,"CvCameraLoaded")
        mOpenCvCameraView.visibility = SurfaceView.VISIBLE
        mOpenCvCameraView.setCameraIndex(1)
        mOpenCvCameraView.setCvCameraViewListener(this)
        mOpenCvCameraView.setCameraPermissionGranted()

    }

    //external fun stringFromJNI(): String

    companion object {
        // Used to load the 'cvcamera' library on application startup.
        init {
            System.loadLibrary("cvcamera")
            System.loadLibrary("opencv_java4")
        }
    }



    override fun onCameraViewStarted(width: Int, height: Int) {
        mRGBA = Mat(height,width,CvType.CV_8UC4)
        mRGBAT = Mat()

    }

    override fun onCameraViewStopped() {
        mRGBA.release()
        mRGBAT.release()
    }

    override fun onCameraFrame(inputFrame: CameraBridgeViewBase.CvCameraViewFrame?): Mat {

        if (inputFrame != null) {
            mRGBA = inputFrame.rgba()
        }
        Core.flip(mRGBA,mRGBAT,-1)
        Imgproc.resize(mRGBAT, mRGBAT, mRGBA.size())
        return mRGBA
    }

    override fun onPointerCaptureChanged(hasCapture: Boolean) {
        super.onPointerCaptureChanged(hasCapture)
    }

    override fun onDestroy() {
        super.onDestroy()
        mOpenCvCameraView.disableView()
    }

    override fun onPause() {
        super.onPause()
        mOpenCvCameraView.disableView()
    }

    override fun onResume() {
        super.onResume()

        if (OpenCVLoader.initDebug()) {
            Log.d(TAG, "OpenCV loaded")
            mLoaderCallback.onManagerConnected(BaseLoaderCallback.SUCCESS)
        }
        else
        {
            Log.d(TAG, "OpenCV didn't load")
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION,this,mLoaderCallback)
        }
        //mOpenCvCameraView.enableView()
    }
}
